### ISSUE TITLE
App updates broken when running NextCloud on OpenWrt

### ISSUE THREAD

**Issue by flyn-org** [nextcloud/server#24136]
I run NextCloud 19.0.4 on OpenWrt. OpenWrt's libcurl does not have support for compression, and I suspect this is getting in the way of NextCloud updating its apps. Here are the libraries OpenWrt's libcurl depends on:

```
# ldd /usr/lib/libcurl.so.4
	ldd (0x7fdb98182000)
	libwolfssl.so.24 => /usr/lib/libwolfssl.so.24 (0x7fdb97fd0000)
	libgcc_s.so.1 => /lib/libgcc_s.so.1 (0x7fdb97fbc000)
	libc.so => ldd (0x7fdb98182000)
```

Loading https://www.flyn.org/share/index.php/settings/apps does not present me with an option to update out-of-date apps. Rather, I see this in the logs:

```
{"reqId":"[ID]","level":2,"time":"2020-11-14T17:12:03+00:00","remoteAddr":"[IP]","user":"[USER]","app":"appstoreFetcher","method":"GET","url":"/share/index.php/settings/apps/list","message":{"Exception":"GuzzleHttp\\Exception\\RequestException","Message":"cURL error 61: Unrecognized content encoding type. libcurl understands identity content encodings. (see https://curl.haxx.se/libcurl/c/libcurl-errors.html)", ... "args":["get","https://apps.nextcloud.com/api/v1/apps.json" ...
```

**Comment by flyn-org** [comment]
I confirmed that compiling libcurl on OpenWrt with libz support works around this bug. Is the HTTP compression scheme negotiation in NextCloud, Guzzle, or the app server broken?

**Comment by LorbusChris** [comment]
It looks like this is the culprit: https://github.com/nextcloud/server/blob/master/lib/private/Http/Client/Client.php#L95 - the only understood compression type is `identity` for curl on OpenWrt, but the linked line will make it so that it requests `gzip`-encoded responses anyway.

Maybe there's a way to override this default accept-encoded header setting, changing it from `gzip` to `identity`.

An even better solution would be to build curl on openwrt with support for compression...

**Comment by LorbusChris** [comment]
I can confirm that changing it from `gzip` to `identity` fixed the issue..

xref: https://gitlab.nic.cz/turris/turris-os-packages/-/issues/662

**Comment by LorbusChris** [comment]
Hi @MorrisJobke, I'm tagging you since you made the change: https://github.com/nextcloud/server/blame/86678ba5df7b74fab9ac739ca17109cd717f8062/lib/private/Http/Client/Client.php#L95 .

tl;dr on OpenWRT curl doesn't support compression, and thus what we need here is a way to configure an override for that hardcoded `gzip` default Accept-Encoding header, replacing it with `identity`.

**Comment by MorrisJobke** [comment]
@LorbusChris @flyn-org Thanks for the report and the deep dive into the issue. Unfortunately I need to say that adding more config options also adds to the complexity of Nextcloud.

Also running Nextcloud on OpenWRT seems to be a pretty rare setup. Thus I would recommend you to just patch out the line in your installation.

Sorry for the inconvenience here, but we have to also use our limited resources that they are useful for as many people as possible.

**Comment by Arusekk** [comment]
I use a Turris router that has Nextcloud out-of-the-box, but even though the system package patches this out, it keeps happening, because of Nextcloud auto-updates; the system disables auto-updates normally, but it is enough to get them wrong just once, because from then on, Nextcloud will not use the patched system package, but the downloaded vanilla upstream code.

I think Nextcloud is auto-updating pretty aggressively, and whenever it updates, the patch is lost. A workaround is to manually edit the offending file every time Nextcloud auto-updates, which is a little bit annoying.

I think the solution is not an extra config knob indeed. But is there no flag to just advertise all supported compression in libcurl, instead of specifically naming one in particular (like `curl --compressed`)? If one day gzip is widely replaced by something different, zstd for example, Nextcloud would transparently benefit. It is hard to believe honestly that such a major feature could be missing, libcurl is such a well-thought piece of software.
